pinot-controller/src/main/java/com/linkedin/pinot/controller/validation/ValidationManager.java

```diff
@@ -137,26 +148,53 @@ public void runValidation() {
       TableType tableType = TableNameBuilder.getTableTypeFromTableName(tableName);
       // For each table, fetch the metadata for all its segments
       if (tableType.equals(TableType.OFFLINE)) {
+        offlineTableCount++;
         validateOfflineSegmentPush(propertyStore, tableName, segmentMetadataList);
+        OfflineTableConfig offlineTableConfig = (OfflineTableConfig) ZKMetadataProvider
+            .getOfflineTableConfig(propertyStore, tableName);
+
+        // To get an approximate size of the IDEALSTATE znode,
+        int replicaCount = offlineTableConfig.getValidationConfig().getReplicationNumber();
+        int instanceCharCount = getServerInstanceCharCount(tableName);
+        long idealstateZnodeSize = computeIdealStateSize(segmentMetadataList, replicaCount, instanceCharCount);
+        _validationMetrics.updateIdealstateZnodeSizeGauge(tableName, idealstateZnodeSize);
       } else if (tableType.equals(TableType.REALTIME)) {
+        realTimeTableCount++;
         List<RealtimeSegmentZKMetadata> realtimeSegmentZKMetadatas = ZKMetadataProvider.getRealtimeSegmentZKMetadataListForTable(propertyStore, tableName);
         boolean countHLCSegments = true;  // false if this table has ONLY LLC segments (i.e. fully migrated)
         AbstractTableConfig tableConfig;
         KafkaStreamMetadata streamMetadata;
+        int replicaCount = 0;
         try {
           tableConfig = _pinotHelixResourceManager.getRealtimeTableConfig(tableName);
           streamMetadata = new KafkaStreamMetadata(tableConfig.getIndexingConfig().getStreamConfigs());
           if (streamMetadata.hasSimpleKafkaConsumerType() && !streamMetadata.hasHighLevelKafkaConsumerType()) {
             countHLCSegments = false;
           }
+          if (streamMetadata.hasSimpleKafkaConsumerType()) {
+            replicaCount = Integer.valueOf(tableConfig.getValidationConfig().getReplicasPerPartition());
+          }
         } catch (Exception e) {
           // TableConfig not present. skip it.
           LOGGER.warn("Cannot get tableconfig for {}", tableName);
           continue;
         }
+        int instanceCharCount = getServerInstanceCharCount(tableName);
+        int llcSegmentCount = 0;
+        long llcSegmentNameCharCount = 0;
         for (RealtimeSegmentZKMetadata realtimeSegmentZKMetadata : realtimeSegmentZKMetadatas) {
           SegmentMetadata segmentMetadata = new SegmentMetadataImpl(realtimeSegmentZKMetadata);
           segmentMetadataList.add(segmentMetadata);
+          if (SegmentName.isLowLevelConsumerSegmentName(segmentMetadata.getName())) {
+            llcSegmentCount++;
+            if (llcSegmentNameCharCount == 0) {
+              llcSegmentNameCharCount = segmentMetadata.getName().length();
+            }
+          }
+        }
+        if (llcSegmentCount > 0) {
```

Comment:
Ah, I see. It does seem more logical to add that to the segment checker (since as you point out, it already fetches IS and EV), though I'm not sure why that's done in a different pass from the validation.